src/soter/openssl/soter_rsa_key_3.c

```diff
@@ -0,0 +1,850 @@
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    /* Validate tag */
+    if (memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)) != 0) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (SOTER_SUCCESS != soter_verify_container_checksum(key)) {
+        return SOTER_DATA_CORRUPT;
+    }
+
+    switch (key->tag[3]) {
+    case RSA_SIZE_TAG_1024:
+        rsa_mod_size = 128;
+        break;
+    case RSA_SIZE_TAG_2048:
+        rsa_mod_size = 256;
+        break;
+    case RSA_SIZE_TAG_4096:
+        rsa_mod_size = 512;
+        break;
+    case RSA_SIZE_TAG_8192:
+        rsa_mod_size = 1024;
+        break;
+    default:
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    if (key_length < rsa_priv_key_size(rsa_mod_size)) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    pub_exp = (const uint32_t*)(curr_bn + ((rsa_mod_size * 4) + (rsa_mod_size / 2)));
+    ;
+    switch (be32toh(*pub_exp)) {
+    case RSA_3:
+    case RSA_F4:
+        break;
+    default:
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    rsa_e = BN_new();
+    rsa_d = BN_new();
+    rsa_p = BN_new();
+    rsa_q = BN_new();
+    rsa_dmp1 = BN_new();
+    rsa_dmq1 = BN_new();
+    rsa_iqmp = BN_new();
+    rsa_n = BN_new();
+    if (!rsa_e || !rsa_d || !rsa_p || !rsa_q || !rsa_dmp1 || !rsa_dmq1 || !rsa_iqmp || !rsa_n) {
+        err = SOTER_NO_MEMORY;
+        goto free_exponents;
+    }
+
+    if (!BN_set_word(rsa_e, be32toh(*pub_exp))) {
+        goto free_exponents;
+    }
+
+    /* Private exponent */
+    if (!BN_bin2bn(curr_bn, rsa_mod_size, rsa_d)) {
+        goto free_exponents;
+    }
+    curr_bn += rsa_mod_size;
+
+    if (!BN_bin2bn(curr_bn, rsa_mod_size / 2, rsa_p)) {
+        goto free_exponents;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* q */
+    if (!BN_bin2bn(curr_bn, rsa_mod_size / 2, rsa_q)) {
+        goto free_exponents;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* dp */
+    if (!BN_bin2bn(curr_bn, rsa_mod_size / 2, rsa_dmp1)) {
+        goto free_exponents;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* dq */
+    if (!BN_bin2bn(curr_bn, rsa_mod_size / 2, rsa_dmq1)) {
+        goto free_exponents;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* qp */
+    if (!BN_bin2bn(curr_bn, rsa_mod_size / 2, rsa_iqmp)) {
+        goto free_exponents;
+    }
+    curr_bn += rsa_mod_size / 2;
+
+    /* modulus */
+    if (!BN_bin2bn(curr_bn, rsa_mod_size, rsa_n)) {
+        goto free_exponents;
+    }
+
+    /* If at least one CRT parameter is zero, free them */
+    if (BN_is_zero(rsa_p) || BN_is_zero(rsa_q) || BN_is_zero(rsa_dmp1) || BN_is_zero(rsa_dmq1)
+        || BN_is_zero(rsa_iqmp)) {
+        BN_free(rsa_p);
+        rsa_p = NULL;
+
+        BN_free(rsa_q);
+        rsa_q = NULL;
+
+        BN_free(rsa_dmp1);
+        rsa_dmp1 = NULL;
+
+        BN_free(rsa_dmq1);
+        rsa_dmq1 = NULL;
+
+        BN_free(rsa_iqmp);
+        rsa_iqmp = NULL;
+    }
+
+    bld = OSSL_PARAM_BLD_new();
```

Comment:
start of the rsa configuration symphony)

Reply:
This makes me think whether `EVP_PKEY_todata()` could be used in RSA private key serialization. Since it got a lot of params being extracted, and if `_todata()` could do it in less number of function calls, it may be more convenient than current approach. Public key on the other hand only got two properties. And EC keys only have one property being read